src/main/java/org/mariadb/jdbc/internal/failover/impl/MastersSlavesListener.java

```diff
@@ -434,80 +434,88 @@ public void lockAndSwitchSecondary(Protocol newSecondaryProtocol) throws QueryEx
      */
     @Override
     public void switchReadOnlyConnection(Boolean mustBeReadOnly) throws QueryException {
-        if (mustBeReadOnly != currentReadOnlyAsked.get() && currentProtocol.inTransaction()) {
+        if (mustBeReadOnly != currentReadOnlyAsked && currentProtocol.inTransaction()) {
             throw new QueryException("Trying to set to read-only mode during a transaction");
         }
         checkWaitingConnection();
-        if (currentReadOnlyAsked.compareAndSet(!mustBeReadOnly, mustBeReadOnly)) {
```

Comment:
@basilmusa The reason for using a volatile is to avoid the overhead of synchronization during the read operations.  If we were not using volatile, even if we used a synchronized block to guard the writes, we would need to synchronize for the reads.  If we did not synchronize on those reads we may see an old version (ie a version on the local CPU cache if no context switches occurred).  It's unlikely, but that is how the java memory model is defined.

If you want references for this, I can dig them up.  But that's was the reason.